src/sentry/api/endpoints/custom_rules.py

```diff
@@ -0,0 +1,157 @@
+            if period is None:
+                raise serializers.ValidationError("Invalid period")
+            if period > MAX_RULE_PERIOD:
+                # limit the expiry period
+                data["period"] = MAX_RULE_PERIOD_STRING
+        return data
+
+
+@region_silo_endpoint
+class CustomRulesEndpoint(OrganizationEndpoint):
+    owner = ApiOwner.TELEMETRY_EXPERIENCE
+
+    publish_status = {
+        "POST": ApiPublishStatus.EXPERIMENTAL,
+    }
+
+    def post(self, request: Request, organization: Organization) -> Response:
+
+        if not features.has("organizations:investigation-bias", organization, actor=request.user):
+            return Response(status=404)
+
+        serializer = CustomRulesInputSerializer(data=request.data)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=400)
+
+        query = serializer.validated_data["query"]
+        projects = serializer.validated_data.get("projects")
+        override_existing = serializer.validated_data.get("overrideExisting")
+        period = serializer.validated_data.get("period")
+
+        try:
+            tokens = parse_search_query(query)
+        except InvalidSearchQuery as e:
+            return Response({"query": [str(e)]}, status=400)
+
+        try:
+            converter = SearchQueryConverter(tokens)
+            condition = converter.convert()
+
+            # the parsing must succeed (it passed validation)
+            delta = cast(timedelta, parse_stats_period(period))
+            expiration = (datetime.utcnow() + delta).timestamp()
+
+            rule = SerializedRule(
+                condition=condition,
+                expiration=expiration,
+                project_ids=projects,
+                org_id=organization.id,
+            )
+
+            _save_rule(rule, override_existing)
+
+        except RuleExistsError:
+            return Response({"query": ["Rule already exists"]}, status=409)
+
+        except ValueError as e:
+            return Response({"query": ["Could not convert to rule", str(e)]}, status=400)
+
+        return Response(rule, status=200)
+
+
+def _save_rule(rule: SerializedRule, override_existing: bool):
+    """
+    Saves a custom rule in the Redis hash for the organization
+    """
+
+    key = custom_rules_redis_key(rule["org_id"])
+    hash_val = get_rule_hash(rule)
+
+    redis_client = get_redis_client_for_ds()
+
+    if rule["expiration"] < datetime.utcnow().timestamp():
+        raise ValueError("Rule has already expired")
+
+    if not override_existing:
+        value = redis_client.hget(key, hash_val)
+        if value is not None:
+            existing_rule = rule_from_json(value)
+            if existing_rule["expiration"] > datetime.utcnow().timestamp():
+                # we have an existing rule that hasn't expired give up
+                raise RuleExistsError()
+
+    redis_client.hset(key, hash_val, json.dumps(rule))
```

Comment:
I would set the expiration of the key as the expiration of the `rule`+ some buffer, in order to make sure that Redis is cleaned up.

Reply:
A key takes a hash of different rules with different expiration times.
I was planning (but didn't get around to do it) to always update the key expiration to be the largest of the individual rules expiration ( plus a little).

This is quite an unfortunate situation ( mostly because I need the